web/packages/teleterm/src/mainProcess/clusterStore/clusterStore.ts

```diff
@@ -0,0 +1,180 @@
+
+import {
+  Cluster,
+  ShowResources,
+} from 'gen-proto-ts/teleport/lib/teleterm/v1/cluster_pb';
+
+import Logger from 'teleterm/logger';
+import { TshdClient } from 'teleterm/services/tshd';
+import { ClusterUri, RootClusterUri, routing } from 'teleterm/ui/uri';
+
+import { AwaitableSender } from '../awaitableSender';
+
+export type State = ReadonlyMap<ClusterUri, Cluster>;
+
+export type ClusterStoreUpdate =
+  /**
+   * Patches allow the other side to keep reference stability
+   * so that not the whole state is recreated.
+   */
+  | { kind: 'patches'; value: Patch[] }
+  /** The full state, useful to get the initial state on the other side. */
+  | { kind: 'state'; value: State };
+
+enablePatches();
+
+export class ClusterStore {
+  private senders = new Set<AwaitableSender<ClusterStoreUpdate>>();
+  private state: State = new Map();
+  private logger = new Logger('ClusterStore');
+
+  constructor(private readonly tshdClient: TshdClient) {}
+
+  /** Adds a cluster. */
+  async add(proxyAddress: string): Promise<Cluster> {
+    const { response } = await this.tshdClient.addCluster({
+      name: proxyAddress,
+    });
+
+    await this.update(draft => {
+      // Do not overwrite the existing cluster;
+      // otherwise we may lose properties fetched from the auth server.
+      // Consider separating properties read from profile and those
+      // fetched from the auth server at the RPC message level.
+      if (draft.has(response.uri)) {
+        return;
+      }
+      draft.set(response.uri, response);
+    });
+
+    return response;
+  }
+
+  /** Logs out of the cluster and removes its profile.*/
+  async logout(uri: RootClusterUri): Promise<void> {
+    // TODO(gzdunek): logout and removeCluster should be combined into
+    //  a single acton in tshd.
+    await this.tshdClient.logout({ clusterUri: uri });
+    await this.tshdClient.removeCluster({ clusterUri: uri });
+    await this.update(draft => {
+      for (let d of draft.values()) {
+        if (routing.belongsToProfile(uri, d.uri)) {
+          draft.delete(d.uri);
+        }
+      }
+    });
+  }
+
+  /**
+   * Synchronizes the root clusters.
+   * Does not make a network call, only reads profiles from the disk.
+   */
+  async syncRootClusters(): Promise<void> {
+    const { response } = await this.tshdClient.listRootClusters({});
+    await this.update(draft => {
+      draft.clear();
+      response.clusters.forEach(cluster => {
+        draft.set(cluster.uri, cluster);
+      });
+    });
+  }
+
+  /**
+   * Synchronizes the root cluster.
+   * Makes a network call to get cluster details and its leaf clusters.
+   */
+  async sync(uri: RootClusterUri): Promise<void> {
+    try {
+      const { cluster, leafs } = await getClusterAndLeafs(this.tshdClient, uri);
+
+      await this.update(draft => {
+        draft.set(cluster.uri, cluster);
+        leafs.forEach(leaf => {
+          draft.set(leaf.uri, leaf);
+        });
+      });
+    } catch (error) {
```

Comment:
Should try/catch wrap only `getClusterAndLeafs` here? The behavior in the catch is something that I think we want to have only if we fail to fetch the cluster.